### ISSUE TITLE
Segmentation fault :SSL_clear

### ISSUE THREAD

**Issue by HoneyryderChuck** [cotag/ruby-tls#11]
I'm reporting this segmentation fault which happens ocasionally in my environment, but happens quite often when ran in a docker image. Here's a stacktrace I pulled from a run in ruby 2.2:

```
/usr/local/bundle/gems/ruby-tls-2.3.1/lib/ruby-tls/ssl.rb:699: [BUG] Segmentation fault at 0x007f8f58000398
ruby 2.2.8p477 (2017-09-14 revision 59906) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0017 p:---- s:0064 e:000063 CFUNC  :SSL_clear
c:0016 p:0086 s:0060 e:000059 METHOD /usr/local/bundle/gems/ruby-tls-2.3.1/lib/ruby-tls/ssl.rb:699
....
```

This seems to happen when calling `#cleanup`, and it looks like a "double-free" kind of situation (cleanup is called many times). I've implemented a workaround to not call cleanup when it's ready:

```ruby
module RubyTls::SSL
	class Box
		attr_reader :ready
	end
end
...
@tls.cleanup if @tls.ready
```

If you think that this workaround is a good solution, I guess you could just apply the same logic inside `#cleanup` (i.e. return early if not `@ready`). However, a solution could also be to refrain from using `SSL_clear`. I'm putting this possibility because [ruby-openssl apparently doesn't call it](https://github.com/ruby/openssl/blob/e72d960db2623b21ee001b5a7b9d9e6ff55bdf94/ext/openssl/ossl_ssl.c#L1902-L1908), and I'm assuming that it might be an avoidable call (but I might be wrong here, haven't checked the docs yet).

**Comment by stakach** [comment]
Thanks for reporting this - checking the ready flag is probably the perfect solution.
I'll update the clear function to check this

**Comment by stakach** [comment]
Fix available in version 2.3.2

**Comment by HoneyryderChuck** [comment]
Sweet! So the `SSL_clear` is necessary, right? I ask because I saw those comments in the `ruby-openssl` source about close being unreliable, and thought that this might be due to that.
